src/zope/interface/registry.py

```diff
@@ -38,6 +41,129 @@ def notify(*arg, **kw): pass
 from zope.interface._compat import STRING_TYPES
 
 
+class _UnhashableComponentCounter(object):
+    # defaultdict(int)-like object for unhashable components
+
+    def __init__(self, otherdict):
+        # [(component, count)]
+        self._data = [item for item in otherdict.items()]
+
+    def __getitem__(self, key):
+        for component, count in self._data:
+            if component == key:
+                return count
+        return 0
+
+    def __setitem__(self, component, count):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                self._data[i] = component, count
+                return
+        self._data.append((component, count))
+
+    def __delitem__(self, component):
+        for i, data in enumerate(self._data):
+            if data[0] == component:
+                del self._data[i]
+                return
+        raise KeyError(component) # pragma: no cover
+
+
+class _UtilityRegistrations(object):
+
+    _regs_for_components = WeakKeyDictionary()
+
+    @classmethod
+    def for_components(cls, components):
+        # We manage these utility/subscription registrations as associated
+        # objects with a weakref to avoid making any changes to
+        # the pickle format
+        try:
+            regs = cls._regs_for_components[components]
+        except KeyError:
+            regs = None
+        else:
+            # In case the components have been re-initted, clear the cache
+            # (zope.component.testing does this between tests)
+            if (regs._utilities is not components.utilities
+                or regs._utility_registrations is not components._utility_registrations):
+                regs = None
```

Comment:
> This lazy operation is probably what's causing the test problems for shoobx.

Because objects are persisting in memory longer than they previously did? They're not accessed (the first time they *would* be, it notices `__init__` has been called and throws them away) so I wouldn't anticipate any problems with closed Connections or the like. Are the tests using `gc.get_objects()` or otherwise doing "leak" tracking?